### ISSUE TITLE
odd behaviour from 'no-unused-vars'

### ISSUE THREAD

**Issue by yeright** [eslint/eslint#6973]
**What version are you using?**
v3.3.1

**What did you do?**
enabled no-unused-vars (as a warning, but also errors)

**What happened?**

Given the following code (simplified), the variable 'server' is flagged as unused.
```js
Client.prototype.addServer = function (serverConfig) {
	
	var server;

	function fallback () { // our retry strategy has failed, connect to the fallback server;
		server = initaliseServer(serverConfig, serverConfig.fallback); // adding the second param here uses the fallback host
	}

	function initaliseServer (serverConfig, fallbackHost) {

		var server = new Server();
		server.on('error', fallback);
		return server;
	} 
	server = initaliseServer(serverConfig);
};
```
**What did you expect to happen?**

Since the server variable DOES get used on execution, I would not expect this to be flagged.
I appreciate I can refactor to remove the use of this outer-scoped variable and probably will, but this behaviour seems wrong.

**Comment by platinumazure** [comment]
Assignments, by themselves, do not count as uses under the rule (see [docs](http://eslint.org/docs/rules/no-unused-vars)).

By the way, the `var server` within `initialiseServer` is a separate variable declaration which shadows the outer declaration. Since ESLint is a static analysis tool, it does not take into account the fact that the inner server is returned and set to the outer server variable. So the outer declaration only gets an assignment on the last line and isn't used anywhere else.

Two options:

1. Use the initialised server after its creation
1. Write an `/* exported server */` comment, if you want to indicate that `server` is intended to be a global variable and used in a different file.

**Comment by ilyavolodin** [comment]
@yeright Your code (I know it's simplified) can be re-written without any loss of functionality as:
```js

function fallback () { // our retry strategy has failed, connect to the fallback server;
    initaliseServer(serverConfig, serverConfig.fallback); // adding the second param here uses the fallback host
}

function initaliseServer (serverConfig, fallbackHost) {

    var server = new Server();
    server.on('error', fallback);
    return server;
} 
initaliseServer(serverConfig);
```

That's why `server` is marked as `unused`.

**Comment by yeright** [comment]
Thanks for the rapid feedback guys, closing this.  I refactored to remove the outer variable.
Must have missed the line in the docs that mentions an assignment not being a 'use'.
